src/Passes/Source/RuleTransformationPass/RulePass.cpp

```diff
@@ -0,0 +1,863 @@
+        // Deleting constants
+        for (auto& x : schedule_instruction_deletion)
+        {
+            x->eraseFromParent();
+        }
+
+        --depth_;
+
+        return true;
+    }
+
+    bool RuleTransformationPass::isActive(llvm::Value* value) const
+    {
+        return active_pieces_.find(value) != active_pieces_.end();
+    }
+
+    void RuleTransformationPass::runCopyAndExpand(llvm::Module& module, llvm::ModuleAnalysisManager&)
+    {
+        replacements_.clear();
+        // For every instruction in every block, we attempt a match
+        // and replace.
+        for (auto& function : module)
+        {
+            if (function.hasFnAttribute(config_.entryPointAttr()))
+            {
+                runOnFunction(function, [this](llvm::Value* value, DeletableInstructions& modifier) {
+                    return copyAndExpand(value, modifier);
+                });
+            }
+        }
+
+        // Active code detection
+        for (auto& global : module.globals())
+        {
+            active_pieces_.insert(&global);
+            for (auto& op : global.operands())
+            {
+                active_pieces_.insert(op);
+            }
+        }
+
+        for (auto& function : module)
+        {
+            bool is_active = false;
+
+            // Checking if the function is referenced by a global variable
+            for (auto user : function.users())
+            {
+                // If the user is active, then it should be expected that the function is also active
+                if (isActive(user))
+                {
+                    is_active = true;
+                    break;
+                }
+            }
+
+            if (is_active || function.hasFnAttribute(config_.entryPointAttr()))
+            {
+                // Marking function as active
+                active_pieces_.insert(&function);
+
+                // Detecting active code
+                runOnFunction(function, [this](llvm::Value* value, DeletableInstructions& modifier) {
+                    return detectActiveCode(value, modifier);
+                });
+            }
+        }
+
+        applyReplacements();
+    }
+
+    void RuleTransformationPass::applyReplacements()
+    {
+        // Applying all replacements
+
+        std::unordered_set<llvm::Value*> already_removed;
+        for (auto it = replacements_.rbegin(); it != replacements_.rend(); ++it)
+        {
+            auto instr1 = llvm::dyn_cast<llvm::Instruction>(it->first);
+
+            // Checking if by accident the same instruction was added
+            if (already_removed.find(instr1) != already_removed.end())
+            {
+                throw std::runtime_error("Instruction was already removed.");
+            }
+            already_removed.insert(instr1);
+
+            if (instr1 == nullptr)
+            {
+                llvm::outs() << "; WARNING: cannot deal with non-instruction replacements\n";
+                continue;
+            }
```

Comment:
(I'm likely wrong but I would like to double-check with you)

I have an impression that in line 477 the `instr1` can become `nullptr`. Line 484 will insert `nullptr` to `already_removed`. Then line 489 will pass the execution to the subsequent iteration of the loop.
In one of the subsequent iterations, in line 477 the `instr1` can become `nullptr` again. Line 480 will `find()` `nullptr` in `already_removed`, and line 482 will fire an exception.

To summarize, I suspect that lines 486 - 490 should be between 477 and 479.

Reply:
You are right. The only reason why it is not a problem is that both are sanity checks that never occur (i.e. instr1 is never null and instr1 is never inserted twice) if the program is correct.

I've changed the order.